### PR TITLE
Added ENRICO-side Timers

### PR DESCRIPTION
**Pull request by RonRahaman**
This PR adds timers to the ENRICO functions, including those that encapsulate functions from the single-physics simulations.  This allows for a much simpler and more consistent interpretation of runtimes.  

The timings are output at each Picard iteration and are presented like this:

```
[ENRICO]: Timer report at i_timestep = 0 , i_picard = 2
[ENRICO]:   CoupledDriver time (seconds, percent)
[ENRICO]:     init_comms            3.5039e-03      0.0002
[ENRICO]:     init_fluid_mask       1.5638e-04      0.0000
[ENRICO]:     init_densities        5.6398e-04      0.0000
[ENRICO]:     init_heat_source      3.3300e-06      0.0000
[ENRICO]:     init_mappings         1.2129e-01      0.0069
[ENRICO]:     init_tallies          1.8890e-03      0.0001
[ENRICO]:     init_temperatures     6.6600e-04      0.0000
[ENRICO]:     init_volumes          2.2250e-03      0.0001
[ENRICO]:     update_density        5.5037e-02      0.0031
[ENRICO]:     update_heat_source    8.8907e-01      0.0508
[ENRICO]:     update_temperature    1.4742e-02      0.0008
[ENRICO]:   NeutronicsDriver time (seconds, percent)
[ENRICO]:     driver_setup          7.6127e+00      0.4349
[ENRICO]:     init_step             2.2221e-01      0.0127
[ENRICO]:     solve_step            2.9868e+02     17.0613
[ENRICO]:     write_step            3.4861e-02      0.0020
[ENRICO]:     finalize_step         6.6170e-06      0.0000
[ENRICO]:   HeatFluidsDriver time (seconds, percent)
[ENRICO]:     driver_setup          3.3093e+02     18.9035
[ENRICO]:     init_step             4.8130e-06      0.0000
[ENRICO]:     solve_step            1.1114e+03     63.4852
[ENRICO]:     write_step            6.6989e-01      0.0383
[ENRICO]:     finalize_step         0.0000e+00      0.0000
[ENRICO]:   Total time (seconds, percent)
[ENRICO]:     total                 1.7506e+03    100.0000
```
These measurements directly correspond to member functions in `CoupledDriver`, `NeutronicsDriver`, and `HeatFluidsDriver`, with these exceptions:
* `init_comms` times the subcommunicator creation, rank discovery, etc., in the CoupledDriver.  This consists of several function calls split in two sections.
* `driver_setup` times the setup functions of the respective single-physics simulation.  In some cases, the timers may surround a single function (e.g.: `openmc_simulation_init()`) and in others, the timers may surround a few relevant functions.  This is not necessarily clear-cut, so I'm definitely willing to adjust the placement of these timers.

### CODE REVIEW

**Comment by RonRahaman** [comment]
@paulromano The implementation of the timers was based on the timers in OpenMC.  Let me know if you'd like clearer attribution.

File: src/timer.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                          const Comm& comm)
+{
+  std::stringstream msg;
+  msg << "  " << header_name << " time (seconds, percent)";
+  comm.message(msg.str());
+  for (const auto& t : times) {
+    std::stringstream msg;
+    msg << "    " << std::setw(22) << std::left << t.name << std::right << std::scientific
+        << std::setprecision(4) << t.time << "    " << std::setw(8) << std::fixed
+        << std::left << std::right << t.percent * 100.0;
```
Comment (paulromano) [inline src/timer.cpp:78]: Not sure which of this changes to the stream are persistent, but it's always a good idea to restore the state of the stream afterward (see [this](https://stackoverflow.com/a/2273352))

File: src/surrogate_heat_driver.cpp [hunk 2]
```diff
@@ -659,6 +661,8 @@ double SurrogateHeatDriver::fluid_temperature(std::size_t pin, std::size_t axial
 
 void SurrogateHeatDriver::write_step(int timestep, int iteration)
 {
+  // TODO: Timers deadlock here...
```
Comment (paulromano) [inline src/surrogate_heat_driver.cpp:664]: Were you planning on taking care of this as part of this PR?

**Review by paulromano — CHANGES_REQUESTED** [review]

File: src/surrogate_heat_driver.cpp (hunk 2 above)
Comment (RonRahaman) [reply src/surrogate_heat_driver.cpp:664]: I spent some time trying to track down the deadlock, but I didn't have any luck.  Is it okay if I raise an Issue for this and defer it to another PR?

File: src/timer.cpp (hunk 1 above)
Comment (RonRahaman) [reply src/timer.cpp:79]: Fixed in https://github.com/enrico-dev/enrico/pull/145/commits/9acca0a5e9bfb287bcaf77d49cde8d511cbc6093

File: src/surrogate_heat_driver.cpp (hunk 2 above)
Comment (paulromano) [reply src/surrogate_heat_driver.cpp:664]: Sounds good to me

**Review by paulromano — APPROVED** [review]

**Pull request merged**